File: agentic-ai/self-heal.js
// Script: self-heal.js
// Purpose: Agentic self-healing loop for Step 5.
//
// Flow:
//   1. Run the generated Playwright test
//   2. If it passes → nothing to heal, report results
//   3. If it fails → launch a headless browser, navigate to the page,
//      inspect the DOM for flash/alert elements, find the correct selector
//   4. Patch the test file with the healed selector
//   5. Re-run the tests to confirm the fix
//   6. Post the full report (failure + fix + passing results) to Jira SCRUM-1

const { execSync, spawnSync } = require('child_process');
const fs = require('fs');
const path = require('path');
const { chromium } = require('@playwright/test/reporter');

require('dotenv').config({ path: path.join(__dirname, '../.env') });

const TEST_FILE   = path.join(__dirname, '../tests/ai-generated/generated-login.spec.ts');
const RESULTS_FILE = path.join(__dirname, '../tests/ai-generated/test-results/generated-login.json');
const JIRA_KEY    = process.env.JIRA_ISSUE_KEY || 'SCRUM-1';

// ─── Helpers ────────────────────────────────────────────────────────────────

function log(msg) { console.log(`\n[self-heal] ${msg}`); }

function runTests() {
  log('Running Playwright tests...');
  const result = spawnSync('npx', [
    'playwright', 'test',
    'tests/ai-generated/generated-login.spec.ts',
    '--reporter=json',
  ], {
    cwd: path.join(__dirname, '..'),
    env: {
      ...process.env,
      PLAYWRIGHT_JSON_OUTPUT_NAME: 'tests/ai-generated/test-results/generated-login.json',
    },
    encoding: 'utf-8',
  });
  return result.status === 0; // true = all passed
}

function readResults() {
  return JSON.parse(fs.readFileSync(RESULTS_FILE, 'utf-8'));
}

// Walk the suite tree and find the first failed test + its error message
function findFailure(suites = []) {
  for (const suite of suites) {
    for (const spec of suite.specs ?? []) {
      for (const test of spec.tests ?? []) {
        const r = test.results?.[0];
        if (r?.status === 'failed') {
          return {
            title: spec.title,
            error: r.error?.message ?? '',
          };
        }
      }
    }
    const nested = findFailure(suite.suites ?? []);
    if (nested) return nested;
  }
  return null;
}

// Extract the broken selector from the error message or scan the test file
function extractBrokenSelector(errorMsg, testSource) {
  // Playwright errors contain "locator('...')" in the call frame
  const fromError = errorMsg.match(/locator\(['"]([^'"]+)['"]\)/);
  if (fromError) return fromError[1];

  // Fallback: scan the test file for locator() calls
  const fromFile = testSource.match(/locator\(['"]([^'"]+)['"]\)/g);
  if (fromFile) return fromFile[0].match(/locator\(['"]([^'"]+)['"]\)/)[1];

  return null;
}

// ─── DOM Inspector ──────────────────────────────────────────────────────────
// Uses Playwright directly (not the test runner) to load the page and
// inspect what flash/alert classes actually exist after login.

async function findCorrectSelector() {
  log('Launching headless browser to inspect live DOM...');
  const { chromium } = require('playwright');
  const browser = await chromium.launch();
  const page = await browser.newPage();

  await page.goto('https://the-internet.herokuapp.com/login');
  await page.fill('#username', 'tomsmith');
  await page.fill('#password', 'SuperSecretPassword!');
  await page.click('button[type=submit]');
  await page.waitForLoadState('domcontentloaded');

  // Get all classes on every .flash* or #flash element
  const candidates = await page.evaluate(() => {
    const els = document.querySelectorAll('[class*="flash"], [id*="flash"], [class*="alert"]');
    return Array.from(els).map(el => ({
      tag:     el.tagName.toLowerCase(),
      id:      el.id,
      classes: Array.from(el.classList),
      text:    el.textContent?.trim().slice(0, 80),
    }));
  });

  await browser.close();
  log('DOM candidates found:\n' + JSON.stringify(candidates, null, 2));

  // Priority 1: element whose classes contain "success" explicitly
  const bySuccessClass = candidates.find(c =>
    c.classes.some(cls => cls.toLowerCase() === 'success')
  );
  if (bySuccessClass && bySuccessClass.classes.length > 0) {
    return '.' + bySuccessClass.classes.join('.');
  }

  // Priority 2: element whose classes contain "success" as substring
  const bySuccessSubstr = candidates.find(c =>
    c.classes.some(cls => cls.toLowerCase().includes('success'))
  );
  if (bySuccessSubstr && bySuccessSubstr.classes.length > 0) {
    return '.' + bySuccessSubstr.classes.join('.');
  }

  // Priority 3: element whose text signals a successful login
  const success = candidates.find(c =>
    c.text?.toLowerCase().includes('secure area') ||
    c.text?.toLowerCase().includes('logged in')
  );
  if (success) {
    if (success.classes.length > 0) return '.' + success.classes.join('.');
    if (success.id) return `#${success.id}`;
  }

  // Fallback: return the first candidate's classes
  if (candidates[0]?.classes.length > 0) {
    return '.' + candidates[0].classes.join('.');
  }

  return null;
}

// ─── File Patcher ───────────────────────────────────────────────────────────

function patchTestFile(brokenSelector, healedSelector) {
  log(`Patching test file: "${brokenSelector}" → "${healedSelector}"`);
  let source = fs.readFileSync(TEST_FILE, 'utf-8');
  const escaped = brokenSelector.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
  const updated = source.replace(new RegExp(escaped, 'g'), healedSelector);
  if (source === updated) {
    log('WARNING: selector not found in file — patch skipped.');
    return false;
  }
  fs.writeFileSync(TEST_FILE, updated, 'utf-8');
  log('File patched successfully.');
  return true;
}

// ─── Jira Reporter ──────────────────────────────────────────────────────────

function postToJira(commentBody) {
  const { JIRA_BASE_URL, JIRA_EMAIL, JIRA_API_TOKEN } = process.env;
  if (!JIRA_BASE_URL || !JIRA_EMAIL || !JIRA_API_TOKEN) {
    log('No Jira credentials — skipping post.');
    return;
  }
  const auth = Buffer.from(`${JIRA_EMAIL}:${JIRA_API_TOKEN}`).toString('base64');
  const baseUrl = JIRA_BASE_URL.replace(/\/$/, '');
  const url = `${baseUrl}/rest/api/3/issue/${JIRA_KEY}/comment`;

  fetch(url, {
    method: 'POST',
    headers: { 'Authorization': `Basic ${auth}`, 'Content-Type': 'application/json' },
    body: JSON.stringify({
      body: {
        type: 'doc', version: 1,
        content: [{ type: 'paragraph', content: [{ type: 'text', text: commentBody }] }],
      },
    }),
  })
    .then(r => r.json())
    .then(d => log(`✅ Jira comment posted to ${JIRA_KEY}: id=${d.id}`))
    .catch(e => log(`❌ Jira post failed: ${e.message}`));
}

// ─── Main Agentic Loop ───────────────────────────────────────────────────────

(async () => {
  // ── Round 1: initial run ──
  const firstRunPassed = runTests();

  if (firstRunPassed) {
    log('All tests passed on first run — nothing to heal. 🎉');
    const results = readResults();
    postToJira(`🤖 Agentic AI — ${JIRA_KEY}\n\n✅ All tests passed on first run. No self-healing required.`);
    return;
  }

  // ── Failure detected ──
  const results1 = readResults();
  const failure = findFailure(results1.suites);
  log(`Failure detected in: "${failure?.title}"`);
  log(`Error: ${failure?.error?.split('\n')[0]}`);

  const testSource = fs.readFileSync(TEST_FILE, 'utf-8');
  const brokenSelector = extractBrokenSelector(failure?.error ?? '', testSource);
  log(`Broken selector identified: "${brokenSelector}"`);

  if (!brokenSelector) {
    log('Could not identify broken selector — manual intervention needed.');
    process.exit(1);
  }

  // ── DOM inspection → find correct selector ──
  const healedSelector = await findCorrectSelector();
  if (!healedSelector) {
    log('DOM inspection found no matching element — manual intervention needed.');
    process.exit(1);
  }
  log(`Healed selector: "${healedSelector}"`);

  // ── Patch the file ──
  const patched = patchTestFile(brokenSelector, healedSelector);
  if (!patched) process.exit(1);

  // ── Round 2: rerun after fix ──
  log('Re-running tests after patch...');
  const secondRunPassed = runTests();
  const results2 = readResults();

  const tests = [];
  function collectTests(suites = []) {
    for (const suite of suites) {
      for (const spec of suite.specs ?? []) {
        for (const test of spec.tests ?? []) {
          const r = test.results?.[0];
          tests.push({ title: spec.title, status: r?.status, duration: r?.duration ?? 0 });
        }
      }
      collectTests(suite.suites ?? []);
    }
  }
  collectTests(results2.suites);

  const testLines = tests.map(t =>
    `  ${t.status === 'passed' ? '✅' : '❌'} ${t.title} (${t.duration}ms)`
  ).join('\n');

  const commentBody = `
🤖 *Agentic Self-Healing Report* — ${JIRA_KEY}

*Broken selector detected:* \`${brokenSelector}\`
*Healed selector applied:*  \`${healedSelector}\`
*Second run result:* ${secondRunPassed ? '✅ ALL PASSED' : '❌ STILL FAILING'}

*Test Results after fix:*
${testLines}

_Self-healed automatically by the agentic CI workflow._
`.trim();

  console.log('\n─── Self-Heal Report ───────────────────────────────────────────────────\n');
  console.log(commentBody);
  console.log('\n────────────────────────────────────────────────────────────────────────\n');

  postToJira(commentBody);

  if (!secondRunPassed) {
    log('Tests still failing after patch — review manually.');
    process.exit(1);
  }

  log('Self-healing complete. Tests green. Jira updated. ✅');
})();

File: agentic-ai/self_heal.py
"""
Script: self_heal.py
Purpose: Agentic self-healing loop.

Flow:
  1. Run the generated Playwright test
  2. If it passes → nothing to heal, report results
  3. If it fails → launch a headless browser, navigate to the page,
     inspect the DOM for flash/alert elements, find the correct selector
  4. Patch the test file with the healed selector
  5. Re-run the tests to confirm the fix
  6. Post the full report (failure + fix + passing results) to Jira

Usage:
    python agentic-ai/self_heal.py

Required .env vars (for Jira reporting): JIRA_BASE_URL, JIRA_EMAIL, JIRA_API_TOKEN
Optional:                                JIRA_ISSUE_KEY (default: SCRUM-1)
"""

import asyncio
import base64
import json
import os
import re
import ssl
import subprocess
import sys
import urllib.request
import urllib.error
from pathlib import Path

from dotenv import load_dotenv

load_dotenv(Path(__file__).parent.parent / ".env")

ROOT      = Path(__file__).parent.parent
JIRA_KEY  = (
    sys.argv[1] if len(sys.argv) > 1
    else os.getenv("STORY_KEY", os.getenv("JIRA_ISSUE_KEY", "SCRUM-1"))
).strip().upper()
TEST_FILE    = ROOT / "tests" / "ai-generated" / f"{JIRA_KEY}.spec.ts"
RESULTS_FILE = ROOT / "tests" / "ai-generated" / "test-results" / f"{JIRA_KEY}-results.json"


# ─── Logging ─────────────────────────────────────────────────────────────────

def log(msg: str) -> None:
    print(f"\n[self-heal] {msg}")


# ─── Test Runner ──────────────────────────────────────────────────────────────

def run_tests() -> bool:
    """Run the generated Playwright spec. Returns True if all tests pass."""
    log("Running Playwright tests...")
    result = subprocess.run(
        ["npx", "playwright", "test",
         str(TEST_FILE.relative_to(ROOT)),
         "--reporter=json"],
        cwd=ROOT,
        env={**os.environ,
             "PLAYWRIGHT_JSON_OUTPUT_NAME": str(RESULTS_FILE)},
        capture_output=True,
        text=True,
    )
    return result.returncode == 0


def read_results() -> dict:
    return json.loads(RESULTS_FILE.read_text(encoding="utf-8"))


# ─── Failure Inspector ────────────────────────────────────────────────────────

def find_failure(suites: list) -> dict | None:
    """Walk the Playwright JSON suite tree and return the first failed test."""
    for suite in suites:
        for spec in suite.get("specs", []):
            for test in spec.get("tests", []):
                r = test.get("results", [{}])[0]
                if r.get("status") == "failed":
                    return {
                        "title": spec.get("title", ""),
                        "error": r.get("error", {}).get("message", ""),
                    }
        nested = find_failure(suite.get("suites", []))
        if nested:
            return nested
    return None


def extract_broken_selector(error_msg: str, test_source: str) -> str | None:
    """Extract the broken selector from the Playwright error message or the test file."""
    # Playwright errors include:  locator('...')
    m = re.search(r"locator\(['\"]([^'\"]+)['\"]\)", error_msg)
    if m:
        return m.group(1)
    # Fallback: first locator() call in the test file
    m = re.search(r"locator\(['\"]([^'\"]+)['\"]\)", test_source)
    return m.group(1) if m else None


# ─── DOM Inspector ────────────────────────────────────────────────────────────

async def find_correct_selector() -> str | None:
    """
    Launch a headless Playwright browser, perform a login, then inspect the DOM
    for any flash/alert elements and return the best CSS selector.
    """
    from playwright.async_api import async_playwright

    log("Launching headless browser to inspect live DOM...")
    async with async_playwright() as p:
        browser = await p.chromium.launch()
        page    = await browser.new_page()

        await page.goto("https://the-internet.herokuapp.com/login")
        await page.fill("#username", "tomsmith")
        await page.fill("#password", "SuperSecretPassword!")
        await page.click("button[type=submit]")
        await page.wait_for_load_state("domcontentloaded")

        candidates = await page.evaluate("""() => {
            const els = document.querySelectorAll('[class*="flash"], [id*="flash"], [class*="alert"]');
            return Array.from(els).map(el => ({
                tag:     el.tagName.toLowerCase(),
                id:      el.id,
                classes: Array.from(el.classList),
                text:    (el.textContent || '').trim().slice(0, 80),
            }));
        }""")

        await browser.close()

    log(f"DOM candidates found:\n{json.dumps(candidates, indent=2)}")

    # Priority 1: class exactly named "success"
    for c in candidates:
        if "success" in [cls.lower() for cls in c["classes"]]:
            return "." + ".".join(c["classes"])

    # Priority 2: class containing "success" as substring
    for c in candidates:
        if any("success" in cls.lower() for cls in c["classes"]):
            return "." + ".".join(c["classes"])

    # Priority 3: text signals successful login
    for c in candidates:
        text_lower = c["text"].lower()
        if "secure area" in text_lower or "logged in" in text_lower:
            if c["classes"]:
                return "." + ".".join(c["classes"])
            if c["id"]:
                return f"#{c['id']}"

    # Fallback: first candidate
    if candidates and candidates[0]["classes"]:
        return "." + ".".join(candidates[0]["classes"])

    return None


# ─── File Patcher ─────────────────────────────────────────────────────────────

def patch_test_file(broken: str, healed: str) -> bool:
    log(f'Patching test file: "{broken}" → "{healed}"')
    source  = TEST_FILE.read_text(encoding="utf-8")
    updated = source.replace(broken, healed)
    if source == updated:
        log("WARNING: selector not found in file — patch skipped.")
        return False
    TEST_FILE.write_text(updated, encoding="utf-8")
    log("File patched successfully.")
    return True


# ─── Jira Reporter ────────────────────────────────────────────────────────────

def post_to_jira(comment_body: str) -> None:
    jira_url   = os.getenv("JIRA_BASE_URL", "").rstrip("/")
    jira_email = os.getenv("JIRA_EMAIL", "")
    jira_token = os.getenv("JIRA_API_TOKEN", "")

    if not all([jira_url, jira_email, jira_token]):
        log("No Jira credentials set — skipping post.")
        return

    auth    = base64.b64encode(f"{jira_email}:{jira_token}".encode()).decode()
    url     = f"{jira_url}/rest/api/3/issue/{JIRA_KEY}/comment"
    payload = json.dumps({
        "body": {
            "type": "doc", "version": 1,
            "content": [{"type": "paragraph", "content": [{"type": "text", "text": comment_body}]}],
        }
    }).encode()
    ctx = ssl.create_default_context()
    ctx.check_hostname = False
    ctx.verify_mode = ssl.CERT_NONE
    req = urllib.request.Request(
        url, data=payload,
        headers={"Authorization": f"Basic {auth}", "Content-Type": "application/json"},
        method="POST",
    )
    try:
        with urllib.request.urlopen(req, timeout=15, context=ctx) as resp:
            data = json.loads(resp.read().decode())
        log(f"✅ Jira comment posted to {JIRA_KEY}: id={data.get('id')}")
    except urllib.error.HTTPError as exc:
        log(f"❌ Jira post failed: {exc.code} {exc.reason}")


# ─── Main Agentic Loop ────────────────────────────────────────────────────────

async def main() -> None:
    # ── Round 1: initial run ──
    first_run_passed = run_tests()

    if first_run_passed:
        log("All tests passed on first run — nothing to heal. 🎉")
        post_to_jira(
            f"🤖 Agentic AI — {JIRA_KEY}\n\n"
            "✅ All tests passed on first run. No self-healing required."
        )
        return

    # ── Failure detected ──
    results1  = read_results()
    failure   = find_failure(results1.get("suites", []))
    log(f"Failure detected in: \"{failure['title']}\"")
    log(f"Error: {failure['error'].splitlines()[0]}")

    test_source     = TEST_FILE.read_text(encoding="utf-8")
    broken_selector = extract_broken_selector(failure["error"], test_source)
    log(f"Broken selector identified: \"{broken_selector}\"")

    if not broken_selector:
        log("Could not identify broken selector — manual intervention needed.")
        sys.exit(1)

    # ── DOM inspection → find correct selector ──
    healed_selector = await find_correct_selector()
    if not healed_selector:
        log("DOM inspection found no matching element — manual intervention needed.")
        sys.exit(1)
    log(f"Healed selector: \"{healed_selector}\"")

    # ── Patch the file ──
    if not patch_test_file(broken_selector, healed_selector):
        sys.exit(1)

    # ── Round 2: rerun after fix ──
    log("Re-running tests after patch...")
    second_run_passed = run_tests()
    results2          = read_results()

    def collect_tests(suites: list) -> list:
        out = []
        for suite in suites:
            for spec in suite.get("specs", []):
                for test in spec.get("tests", []):
                    r = test.get("results", [{}])[0]
                    out.append({
                        "title":    spec.get("title", ""),
                        "status":   r.get("status", "unknown"),
                        "duration": r.get("duration", 0),
                    })
            out.extend(collect_tests(suite.get("suites", [])))
        return out

    tests      = collect_tests(results2.get("suites", []))
    test_lines = "\n".join(
        f"  {'✅' if t['status'] == 'passed' else '❌'} {t['title']} ({t['duration']}ms)"
        for t in tests
    )

    comment_body = (
        f"🤖 *Agentic Self-Healing Report* — {JIRA_KEY}\n\n"
        f"*Broken selector detected:* `{broken_selector}`\n"
        f"*Healed selector applied:*  `{healed_selector}`\n"
        f"*Second run result:* {'✅ ALL PASSED' if second_run_passed else '❌ STILL FAILING'}\n\n"
        f"*Test Results after fix:*\n{test_lines}\n\n"
        "_Self-healed automatically by the agentic CI workflow._"
    )

    print("\n─── Self-Heal Report ───────────────────────────────────────────────────\n")
    print(comment_body)
    print("\n────────────────────────────────────────────────────────────────────────\n")

    post_to_jira(comment_body)

    if not second_run_passed:
        log("Tests still failing after patch — review manually.")
        sys.exit(1)

    log("Self-healing complete. Tests green. Jira updated. ✅")


if __name__ == "__main__":
    asyncio.run(main())
